Use table direction, not section direction, for ordering cells and borders.

fantasai opened the pull request:
#### 86dc22fbb4b7f2d718008ecf8545701f98c50cc9
<pre>
Use table direction, not section direction, for ordering cells and borders.
<a href="https://bugs.webkit.org/show_bug.cgi?id=96691">https://bugs.webkit.org/show_bug.cgi?id=96691</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=79272">https://bugs.webkit.org/show_bug.cgi?id=79272</a>
<a href="https://rdar.apple.com/problem/99343532">rdar://problem/99343532</a>

Reviewed by Alan Baradlay.

This commit switches table layout to use the direction of the table,
not the section.

In order to make this readable, it updates RenderStyle&apos;s border accessors
to accept an optional style argument, from which it takes the writing mode
for directional conversion.

Co-Authored-By: Julian Chaffraix &lt;jchaffraix@webkit.org&gt;

* LayoutTests/TestExpectations:
* LayoutTests/fast/table/border-collapsing/first-cell-left-border-hidden-table-ltr-section-rtl.html:
* LayoutTests/fast/table/border-collapsing/last-cell-left-border-hidden-table-ltr-section-rtl.html:
* LayoutTests/fast/table/paint-section-borders-without-cells-vertical-rl-rtl-expected.html: Added.
* LayoutTests/fast/table/paint-section-borders-without-cells-vertical-rl-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-002.html: Added.
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::tableStartBorderAdjoiningCell const): Deleted.
(WebCore::RenderTable::tableEndBorderAdjoiningCell const): Deleted.
* Source/WebCore/rendering/RenderTable.h:
(WebCore::isDirectionSame): Deleted.
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::hasStartBorderAdjoiningTable const):
(WebCore::RenderTableCell::hasEndBorderAdjoiningTable const):
(WebCore::RenderTableCell::computeCollapsedStartBorder const):
(WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
(WebCore::RenderTableCell::computeCollapsedAfterBorder const):
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCellInlines.h:
(WebCore::RenderTableCell::borderAdjoiningCellAfter):
(WebCore::RenderTableCell::borderAdjoiningCellBefore):
(WebCore::RenderTableCell::borderAdjoiningTableEnd const):
(WebCore::RenderTableCell::borderAdjoiningTableStart const):
* Source/WebCore/rendering/RenderTableCol.cpp:
(WebCore::RenderTableCol::borderAdjoiningCellStartBorder const):
(WebCore::RenderTableCol::borderAdjoiningCellEndBorder const):
(WebCore::RenderTableCol::borderAdjoiningCellBefore const):
(WebCore::RenderTableCol::borderAdjoiningCellAfter const):
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::borderAdjoiningStartCell const):
(WebCore::RenderTableRow::borderAdjoiningEndCell const):
* Source/WebCore/rendering/RenderTableRowInlines.h:
(WebCore::RenderTableRow::borderAdjoiningTableStart const):
(WebCore::RenderTableRow::borderAdjoiningTableEnd const):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcOuterBorderBefore const):
(WebCore::RenderTableSection::calcOuterBorderAfter const):
(WebCore::RenderTableSection::calcOuterBorderStart const):
(WebCore::RenderTableSection::calcOuterBorderEnd const):
(WebCore::RenderTableSection::logicalRectForWritingModeAndDirection const):
(WebCore::RenderTableSection::offsetLeftForRowGroupBorder):
(WebCore::RenderTableSection::offsetTopForRowGroupBorder):
(WebCore::RenderTableSection::verticalRowGroupBorderHeight):
(WebCore::RenderTableSection::horizontalRowGroupBorderWidth):
(WebCore::RenderTableSection::paintObject):
(WebCore::RenderTableSection::borderAdjoiningStartCell const):
(WebCore::RenderTableSection::borderAdjoiningEndCell const):
(WebCore::RenderTableSection::setLogicalPositionForCell const):
(WebCore::RenderTableSection::firstRowCellAdjoiningTableStart const): Deleted.
(WebCore::RenderTableSection::firstRowCellAdjoiningTableEnd const): Deleted.
* Source/WebCore/rendering/RenderTableSection.h:
* Source/WebCore/rendering/RenderTableSectionInlines.h:
(WebCore::RenderTableSection::borderAdjoiningTableEnd const):
(WebCore::RenderTableSection::borderAdjoiningTableStart const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::borderBefore const):
(WebCore::RenderStyle::borderAfter const):
(WebCore::RenderStyle::borderStart const):
(WebCore::RenderStyle::borderEnd const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::borderAfter const):
(WebCore::RenderStyle::borderBefore const):
(WebCore::RenderStyle::borderEnd const):
(WebCore::RenderStyle::borderStart const):

Canonical link: <a href="https://commits.webkit.org/283875@main">https://commits.webkit.org/283875@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57a4c8203f51f06cd25acb03715360d96f45960b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18508 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54115 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12513 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70634 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43080 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58434 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34614 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39753 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15839 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17062 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61711 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73317 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11526 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61557 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11561 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58503 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61621 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15002 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9414 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3027 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42752 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43828 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43569 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -527,6 +527,7 @@ class RenderBox : public RenderBoxModelObject {
     LayoutUnit lineHeight(bool firstLine, LineDirectionMode, LinePositionMode = PositionOnContainingLine) const override;
     LayoutUnit baselinePosition(FontBaseline, bool firstLine, LineDirectionMode, LinePositionMode = PositionOnContainingLine) const override;
 
+
```
intentional extra line?

fantasai replied:
Fixed. I'll push after layout tests finish running. :)

fantasai committed:
Use table direction, not section direction, for ordering cells and borders.
https://bugs.webkit.org/show_bug.cgi?id=96691
https://bugs.webkit.org/show_bug.cgi?id=79272
rdar://problem/99343532

Reviewed by Alan Baradlay.

This commit switches table layout to use the direction of the table,
not the section.

In order to make this readable, it updates RenderStyle's border accessors
to accept an optional style argument, from which it takes the writing mode
for directional conversion.

Co-Authored-By: Julian Chaffraix <jchaffraix@webkit.org>

* LayoutTests/TestExpectations:
* LayoutTests/fast/table/border-collapsing/first-cell-left-border-hidden-table-ltr-section-rtl.html:
* LayoutTests/fast/table/border-collapsing/last-cell-left-border-hidden-table-ltr-section-rtl.html:
* LayoutTests/fast/table/paint-section-borders-without-cells-vertical-rl-rtl-expected.html: Added.
* LayoutTests/fast/table/paint-section-borders-without-cells-vertical-rl-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/table-progression-htb-002.html: Added.
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::tableStartBorderAdjoiningCell const): Deleted.
(WebCore::RenderTable::tableEndBorderAdjoiningCell const): Deleted.
* Source/WebCore/rendering/RenderTable.h:
(WebCore::isDirectionSame): Deleted.
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::hasStartBorderAdjoiningTable const):
(WebCore::RenderTableCell::hasEndBorderAdjoiningTable const):
(WebCore::RenderTableCell::computeCollapsedStartBorder const):
(WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
(WebCore::RenderTableCell::computeCollapsedAfterBorder const):
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCellInlines.h:
(WebCore::RenderTableCell::borderAdjoiningCellAfter):
(WebCore::RenderTableCell::borderAdjoiningCellBefore):
(WebCore::RenderTableCell::borderAdjoiningTableEnd const):
(WebCore::RenderTableCell::borderAdjoiningTableStart const):
* Source/WebCore/rendering/RenderTableCol.cpp:
(WebCore::RenderTableCol::borderAdjoiningCellStartBorder const):
(WebCore::RenderTableCol::borderAdjoiningCellEndBorder const):
(WebCore::RenderTableCol::borderAdjoiningCellBefore const):
(WebCore::RenderTableCol::borderAdjoiningCellAfter const):
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::borderAdjoiningStartCell const):
(WebCore::RenderTableRow::borderAdjoiningEndCell const):
* Source/WebCore/rendering/RenderTableRowInlines.h:
(WebCore::RenderTableRow::borderAdjoiningTableStart const):
(WebCore::RenderTableRow::borderAdjoiningTableEnd const):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcOuterBorderBefore const):
(WebCore::RenderTableSection::calcOuterBorderAfter const):
(WebCore::RenderTableSection::calcOuterBorderStart const):
(WebCore::RenderTableSection::calcOuterBorderEnd const):
(WebCore::RenderTableSection::logicalRectForWritingModeAndDirection const):
(WebCore::RenderTableSection::offsetLeftForRowGroupBorder):
(WebCore::RenderTableSection::offsetTopForRowGroupBorder):
(WebCore::RenderTableSection::verticalRowGroupBorderHeight):
(WebCore::RenderTableSection::horizontalRowGroupBorderWidth):
(WebCore::RenderTableSection::paintObject):
(WebCore::RenderTableSection::borderAdjoiningStartCell const):
(WebCore::RenderTableSection::borderAdjoiningEndCell const):
(WebCore::RenderTableSection::setLogicalPositionForCell const):
(WebCore::RenderTableSection::firstRowCellAdjoiningTableStart const): Deleted.
(WebCore::RenderTableSection::firstRowCellAdjoiningTableEnd const): Deleted.
* Source/WebCore/rendering/RenderTableSection.h:
* Source/WebCore/rendering/RenderTableSectionInlines.h:
(WebCore::RenderTableSection::borderAdjoiningTableEnd const):
(WebCore::RenderTableSection::borderAdjoiningTableStart const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::borderBefore const):
(WebCore::RenderStyle::borderAfter const):
(WebCore::RenderStyle::borderStart const):
(WebCore::RenderStyle::borderEnd const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::borderAfter const):
(WebCore::RenderStyle::borderBefore const):
(WebCore::RenderStyle::borderEnd const):
(WebCore::RenderStyle::borderStart const):

Canonical link: https://commits.webkit.org/283875@main
